### ISSUE TITLE
Can't add a domain mapping to the shopping cart on an existing site without a plan

### ISSUE THREAD

**Issue by DavidRothstein** [Automattic/wp-calypso#49746]
### Steps to reproduce the behavior

1. Start with an existing site that doesn't have a plan.
2. Go to Upgrades -> Domains -> "Add a domain to this site" -> "Use a domain I own" and try to map a new domain.
3. You get redirected to the Plans page with an empty cart; the domain mapping can't be added.

### What I expected to happen

Although domain mappings aren't allowed on sites without a plan, there is code that is designed to automatically add a plan to the shopping cart if you try to purchase a domain mapping in that scenario.  It's a regression that this is no longer working -- and the current experience is pretty bad.

### Context

Although the code that either clears the cart (current behavior) or automatically adds a plan (intended behavior) lives on the server, it is responding to an `is_update` property sent by Calypso.  The expectation is that this property is set to `true` only when an existing cart is being updated (e.g. removing a product).  But it seems to now be set to `true` for new carts also.  Note there is some other server-side code in the shopping cart that checks the `is_update` property also and that could be affected by this.

To fix the bug, we either need to restore the previous behavior of `is_update` or otherwise separately figure out how to distinguish between removing and adding items to the cart.

### Level of impact

I'm assuming many people add domain mappings during signup (where this bug does not happen) but probably a sizable number do it afterwards.  If they don't already have a plan or can't figure out that they need to add one to the shopping cart manually, they could get pretty stuck on this bug.

cc @Automattic/cobalt also

**Comment by sirbrillig** [comment]
It's not clear from the name but the intent of `is_update` was to differentiate between cases when a cart is being modified during checkout and when it's being modified outside of checkout, specifically to fix one use case with the shopping-cart endpoint. First let me describe the bug as it existed.

Remember that the shopping-cart endpoint has no concept of "adding" or "removing" products; when you call the endpoint you're requesting the creation of a brand-new cart, even if that cart contains many of the same products that it did before the request

There's logic in the cart endpoint which says, "after the cart is created, if it contains a domain product and the customer's site does not have a plan, and there's no plan in the cart, add a Premium plan to the cart". This logic was convenient because a customer could add, say, domain mapping to the cart and they'd end up in checkout with domain mapping and a plan (since we didn't allow domain mapping without a plan).

The bug happened because the same logic would be hit if a customer had added a plan and a domain product to the cart, and then removed the plan from the cart. In that case the endpoint (not knowing that you were "removing" something) would add the Premium plan again, making for a very confusing UX.

This was fixed by adding the `is_update` property in D40211-code which allowed us to update the logic and say, "if you're removing a plan from a cart that has a domain product, then don't bother adding the Premium plan, just remove the domain product also". It was probably the wrong solution because what we really needed to know is not if a cart request was an "update" but if the request was a "removal". The resulting confusion is that, really, all uses of the shopping cart in calypso are "updates"; we never create a cart from scratch unless one does not exist. We could certainly clarify this usage in the code, but I don't think it's a very useful distinction anyway. We may just want to remove it at this point. The only time a cart operation is not `is_update` is when it's a server-to-server operation.

Instead, we can fix the logic in the endpoint as it should have been fixed in the first place, by using the previous version of the cart to determine if we're removing a plan or adding a domain product and then making the appropriate decision if we should add an extra plan to the cart or not. I've started working on this in D56590-code. However, one thing we need to determine is when this logic should be triggered because the rules about when a plan is required for a domain product have changed several times and I don't remember what they are right now. Moreover, the tests for the behavior described above were removed in D41610-code, so in making this change we'll need to understand why they were removed and probably recreate them.
